Mathlib/RingTheory/Flat/FaithfullyFlat.lean

```diff
@@ -163,6 +169,186 @@ lemma iff_flat_and_lTensor_reflects_triviality :
+
+Let `N₁ -l₁₂-> N₂ -l₂₃-> N₃` be two linear maps.
+- We first show that if `N₁ ⊗ M -> N₂ ⊗ M -> N₃ ⊗ M` is exact, then `N₁ -l₁₂-> N₂ -l₂₃-> N₃` is a
+  complex, i.e. `range l₁₂ ≤ ker l₂₃`.
+  This is `range_le_ker_of_exact_rTensor`.
+- Then in `rTensor_reflects_exact`, we show `ker l₂₃ = range l₁₂` by considering the cohomology
+  `ker l₂₃ ⧸ range l₁₂`.
+This shows that when `M` is faithfully flat, `- ⊗ M` reflects exact sequences. For details, see
+comments in the proof. Since `M` is flat, `- ⊗ M` preserves exact sequences.
+
+On the other hand, if `- ⊗ M` preserves and reflects exact sequences, then `M` is faithfully flat.
+- `M` is flat because `- ⊗ M` preserves exact sequences.
+- We need to show that if `N ⊗ M = 0` then `N = 0`. Consider the sequence `N -0-> N -0-> 0`. After
+  tensoring with `M`, we get `N ⊗ M -0-> N ⊗ M -0-> 0` which is exact because `N ⊗ M = 0`.
+  Since `- ⊗ M` reflects exact sequences, `N = 0`.
+-/
+
+section arbitrary_universe
+
+variable ⦃N1 : Type*⦄ [AddCommGroup N1] [Module R N1]
+variable ⦃N2 : Type*⦄ [AddCommGroup N2] [Module R N2]
+variable ⦃N3 : Type*⦄ [AddCommGroup N3] [Module R N3]
+variable (l12 : N1 →ₗ[R] N2) (l23 : N2 →ₗ[R] N3)
+
+/--
+If `M` is faithfully flat, then exactness of `N₁ ⊗ M -> N₂ ⊗ M -> N₃ ⊗ M` implies that the
+composition `N₁ -> N₂ -> N₃` is `0`.
+
+Implementation detail, please use `rTensor_reflects_exact` instead.
+-/
+lemma range_le_ker_of_exact_rTensor [fl : FaithfullyFlat R M]
+    (ex : Function.Exact (l12.rTensor M) (l23.rTensor M)) :
+    LinearMap.range l12 ≤ LinearMap.ker l23 := by
+  -- let `n1 ∈ N1`. We need to show `l23 (l12 n1) = 0`. Suppose this is not the case.
+  rintro _ ⟨n1, rfl⟩
+  show l23 (l12 n1) = 0
+  by_contra! hn1
+  -- Let `E` be the submodule spanned by `l23 (l12 n1)`. Then because `l23 (l12 n1) ≠ 0`, we have
+  -- `E ≠ 0`.
+  let E : Submodule R N3 := Submodule.span R {l23 (l12 n1)}
+  have hE : Nontrivial E :=
+    ⟨0, ⟨⟨l23 (l12 n1), Submodule.mem_span_singleton_self _⟩, Subtype.coe_ne_coe.1 hn1.symm⟩⟩
+
+  -- Since `N1 ⊗ M -> N2 ⊗ M -> N3 ⊗ M` is exact, we have `l23 (l12 n1) ⊗ₜ m = 0` for all `m : M`.
+  have eq1 : ∀ (m : M), l23 (l12 n1) ⊗ₜ[R] m = 0 := fun m ↦
+    ex.apply_apply_eq_zero (n1 ⊗ₜ[R] m)
+  -- Then `E ⊗ M = 0`. Indeed,
+  have eq0 : (⊤ : Submodule R (E ⊗[R] M)) = ⊥ := by
+    -- suppose `x ∈ E ⊗ M`. We will show `x = 0`.
+    ext x
+    simp only [Submodule.mem_top, Submodule.mem_bot, true_iff]
+    have mem : x ∈ (⊤ : Submodule R _) := ⟨⟩
+    rw [← TensorProduct.span_tmul_eq_top, mem_span_set] at mem
+    obtain ⟨c, hc, rfl⟩ := mem
+    choose b a hy using hc
+    let r :  ⦃a : E ⊗[R] M⦄ → a ∈ ↑c.support → R := fun a ha =>
+      Submodule.mem_span_singleton.1 (b ha).2 |>.choose
+    have hr : ∀ ⦃i : E ⊗[R] M⦄ (hi : i ∈ c.support), b hi =
+        r hi • ⟨l23 (l12 n1), Submodule.mem_span_singleton_self _⟩ := by
+      intro i hi
+      ext
+      exact Submodule.mem_span_singleton.1 (b hi).2 |>.choose_spec.symm
```

Comment:
It seems you have more than one goal here, can you use the focusing dot?

Reply:
hmm, I think I only have one goal here

Reply:
I mean that there is an `exact ...` line not indented that close a goal. What happens next?

Reply:
Sorry I still don't get it. The `exact`-statement on line 240 closes the `have`-statement on line 236. I changed the tactic-mode proof to a term-mode one now.

Reply:
I probably just misread the indentation in the github editor. Forget about it.